kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -351,6 +357,26 @@ private Optional<Long> loadLatestTransactionVersion(Engine engine, String applic
     return Optional.empty();
   }
 
+  /**
+   * Loads the domain metadata map, either from CRC info (if available) or from the transaction log.
+   * Note that when loading from CRC info, tombstones (removed domains) are not preserved, while
+   * they are preserved when loading from the transaction log.
+   *
+   * @param engine The engine to use for loading from log when necessary
+   * @return A map of domain names to their metadata
+   */
+  private Map<String, DomainMetadata> loadDomainMetadataMap(Engine engine) {
+    // First try to load from CRC info if available
+    if (currentCrcInfo.isPresent() && currentCrcInfo.get().getDomainMetadata().isPresent()) {
+      return currentCrcInfo.get().getDomainMetadata().get().stream()
+          .collect(Collectors.toMap(DomainMetadata::getDomain, Function.identity()));
+    }
```

Comment:
Qn: if `currentCrcInfo.isPresent()` is true, isn't the expectation that the domain metadata in that CRC is valid and no need to do log replay?

Reply:
Technically domain metadata is an optional field, so we cannot guarantee for all cases, crc has domain metadata. (e.g. delta spark written crc only keep 10 dm by default. If size of dm is larger than 10, it won't persist crc)